poly/src/domain/radix2/fft.rs

```diff
@@ -141,94 +141,160 @@ impl<F: FftField> Radix2EvaluationDomain<F> {
         // In the sequential case, we will keep on making the roots cache-aligned,
         // according to the access pattern that the FFT uses.
         // It is left as a TODO to implement this for the parallel case
-        let roots = &mut self.roots_of_unity(root);
+        let mut roots = self.roots_of_unity(root);
+
+        #[cfg(feature = "parallel")]
+        let max_threads = rayon::current_num_threads();
 
         #[cfg(not(feature = "parallel"))]
-        let mut root_len = roots.len();
+        let max_threads = 1;
 
         let mut gap = xi.len() / 2;
         while gap > 0 {
             // each butterfly cluster uses 2*gap positions
             let chunk_size = 2 * gap;
-            #[cfg(feature = "parallel")]
-            let nchunks = xi.len() / chunk_size;
+            let num_chunks = xi.len() / chunk_size;
+            let sub_chunks = (max_threads - 1) / num_chunks + 1;
+            let sub_chunk_size = chunk_size / (2 * sub_chunks);
 
-            let butterfly_fn = |(chunk_index, (lo, hi)): (usize, (&mut T, &mut T))| {
+            let butterfly_fn = |(index, (lo, hi)): (usize, (&mut T, &mut T))| {
                 let neg = *lo - *hi;
                 *lo += *hi;
 
                 *hi = neg;
 
-                #[cfg(feature = "parallel")]
-                let index = nchunks * chunk_index;
-                // Due to cache aligning, the index into roots is the chunk index
-                #[cfg(not(feature = "parallel"))]
-                let index = chunk_index;
-
                 *hi *= roots[index];
             };
 
-            ark_std::cfg_chunks_mut!(xi, chunk_size).for_each(|cxi| {
+            cfg_chunks_mut!(xi, chunk_size).for_each(|cxi| {
                 let (lo, hi) = cxi.split_at_mut(gap);
                 // If the chunk is sufficiently big that parallelism helps,
                 // we parallelize the butterfly operation within the chunk.
                 //
                 // if chunk_size > MIN_CHUNK_SIZE_FOR_PARALLELIZATION
+
                 if gap > MIN_CHUNK_SIZE_FOR_PARALLELIZATION / 2 {
-                    cfg_iter_mut!(lo).zip(hi).enumerate().for_each(butterfly_fn);
+                    cfg_chunks_mut!(lo, sub_chunk_size)
+                        .zip(cfg_chunks_mut!(hi, sub_chunk_size))
+                        .enumerate()
+                        .for_each(|(chunk_id, (lo_chunk, hi_chunk))| {
+                            lo_chunk
+                                .iter_mut()
+                                .zip(hi_chunk)
+                                .enumerate()
+                                .for_each(|(idx, d)| {
+                                    butterfly_fn((chunk_id * sub_chunk_size + idx, d))
```

Comment:
Let's extract this into a closure that can be used in both branches:

```rust
let butterfly = |lo, hi, index_transformer| {
	lo
		.iter_mut()
    	.zip(hi_chunk)
		.enumerate()
    	.for_each(|(idx, (lo, hi))| {
     	    let neg = *lo - *hi;
    	    *lo += *hi;
	
       		*hi = neg;

     		*hi *= roots[index_transformer(idx)];
     	}
};
```
where `index_transformer = |a| a;` in the serial case, and `index_transformer = |idx| chunk_id * sub_chunk_size + idx;` in the parallel case.

Reply:
Maybe we can even abstract this into a separate helper function that's common to both the `io_helper` and `oi_helper` settings.

Reply:
I don't think this is as straightforward as you think. The code might not be very readable. But the main issue is that it uses a closure defined function. Which is actually bad. I think what should be factored out are the butterfly functions.